rust/src/dirdiff.rs

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright (C) Red Hat, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ */
+
+use anyhow::{Context, Result};
+use openat_ext::OpenatDirExt;
+use serde_derive::{Deserialize, Serialize};
+use std::borrow::Cow;
+use std::collections::BTreeSet;
+use std::convert::TryFrom;
+use std::fmt;
+
+use std::io::Read;
+
+pub(crate) type FileSet = BTreeSet<String>;
+
+/// Diff between two directories.
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub(crate) struct Diff {
+    /// Files that are new in an existing directory
+    pub(crate) added_files: FileSet,
+    /// New directories
+    pub(crate) added_dirs: FileSet,
+    /// Files removed
+    pub(crate) removed_files: FileSet,
+    /// Directories removed (recursively)
+    pub(crate) removed_dirs: FileSet,
+    /// Files that changed (in any way, metadata or content)
+    pub(crate) changed_files: FileSet,
+    /// Directories that changed mode/permissions
+    pub(crate) changed_dirs: FileSet,
+}
+
+impl Diff {
+    #[allow(dead_code)]
+    pub(crate) fn count(&self) -> usize {
+        self.added_files.len()
+            + self.added_dirs.len()
+            + self.removed_files.len()
+            + self.removed_dirs.len()
+            + self.changed_files.len()
+            + self.changed_dirs.len()
+    }
+
+    pub(crate) fn contains(&self, s: &str) -> bool {
+        self.added_files.contains(s)
+            || self.added_dirs.contains(s)
+            || self.removed_files.contains(s)
+            || self.removed_dirs.contains(s)
+            || self.changed_files.contains(s)
+            || self.changed_dirs.contains(s)
+    }
+}
+
+impl fmt::Display for Diff {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "files(added:{} removed:{} changed:{}) dirs(added:{} removed:{} changed:{})",
+            self.added_files.len(),
+            self.removed_files.len(),
+            self.changed_files.len(),
+            self.added_dirs.len(),
+            self.removed_dirs.len(),
+            self.changed_dirs.len()
+        )
+    }
+}
+
+fn file_content_changed(
+    src: &openat::Dir,
+    dest: &openat::Dir,
+    path: &str,
+    expected_len: u64,
+) -> Result<bool> {
```

Comment:
All the `?`s in this block would benefit from having some context, otherwise it gets pretty hard debugging I/O errors.

Reply:
Since these are all reading, not writing, errors are less likely; I think the outer "diff failed: Failed to read" might be enough.  But I agree for sure long term.

But we have a bigger problem right now which is https://github.com/dtolnay/cxx/issues/290#issuecomment-756432907